src/cpu/dyn_cache.h

```diff
@@ -710,12 +723,50 @@ static void cache_block_closing(const uint8_t *block_start, Bitu block_size);
 #endif
 
 /* Define temporary pagesize so the MPROTECT case and the regular case share as much code as possible */
-#if defined(HAVE_MPROTECT)
+#if defined(HAVE_MPROTECT) || defined(HAVE_MMAP)
 #define PAGESIZE_TEMP PAGESIZE
 #else
 #define PAGESIZE_TEMP 4096
 #endif
 
+static constexpr size_t cache_code_start_sz = CACHE_TOTAL + CACHE_MAXSIZE + PAGESIZE_TEMP - 1 + PAGESIZE_TEMP;
+static constexpr size_t cache_blocks_sz = CACHE_BLOCKS * sizeof(CacheBlock);
+
+static inline void dyn_mem_execute()
+{
+#if defined(HAVE_PTHREAD_WRITE_PROTECT_NP)
+    if (__builtin_available(macOS 11.0, *)) {
+        pthread_jit_write_protect_np(true);
+    }
+#elif defined(HAVE_MPROTECT)
+	if (mprotect(cache_code_start_ptr, cache_code_start_sz, PROT_EXEC | PROT_READ) == -1) {
+		E_Exit("mprotect failed in dyn_mem_execute");		
+	}
+#endif
+}
+
+static inline void dyn_mem_write()
+{
+#if defined(HAVE_PTHREAD_WRITE_PROTECT_NP)
+    if (__builtin_available(macOS 11.0, *)) {
+        pthread_jit_write_protect_np(false);
+    }
+#elif defined(HAVE_MPROTECT)
+	if (mprotect(cache_code_start_ptr, cache_code_start_sz, PROT_WRITE | PROT_READ) == -1) {
+		E_Exit("mprotect failed in dyn_mem_write");			
```

Comment:
For macOS systems pre-11.0, is there value in falling back to `mprotect`, or is a noop sufficient?

Reply:
I will need to test this on an older system. The Hardened Runtime shipped with Mojave and I believe that will enforce memory protection, so my initial guess is that an mprotect will be required (it would be nice to ship our macOS binary(ies) with the Hardened Runtime enabled at some point).